bokeh/application/handlers/tests/test_code.py

```diff
@@ -36,69 +34,131 @@ class SomeModelInTestScript(Model):
+    handler = CodeHandler(source="# This script does nothing", filename="/test_filename")
+    handler.modify_document(doc)
+    if handler.failed:
+        raise RuntimeError(handler.error)
+
+    assert not doc.roots
+
+def test_script_adds_roots():
+    doc = Document()
+    handler = CodeHandler(source=script_adds_two_roots, filename="/test_filename")
+    handler.modify_document(doc)
+    if handler.failed:
+        raise RuntimeError(handler.error)
+
+    assert len(doc.roots) == 2
+
+def test_script_bad_syntax():
+    doc = Document()
+    handler = CodeHandler(source="This is a syntax error", filename="/test_filename")
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert 'Invalid syntax' in handler.error
+
+def test_script_runtime_error():
+    doc = Document()
+    handler = CodeHandler(source="raise RuntimeError('nope')", filename="/test_filename")
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert 'nope' in handler.error
+
+def test_script_sys_path():
+    doc = Document()
+    handler = CodeHandler(source="""import sys; raise RuntimeError("path: '%s'" % sys.path[0])""", filename="/test_filename")
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert "path: ''" in handler.error
+
+def test_script_cwd():
+    doc = Document()
+    handler = CodeHandler(source="""import os; raise RuntimeError("cwd: '%s'" % os.getcwd())""", filename="/test_filename")
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert "cwd: '/'" in handler.error
+
+def test_script_argv():
+    doc = Document()
+    handler = CodeHandler(source="""import sys; raise RuntimeError("argv: %r" % sys.argv)""", filename="/test_filename")
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert "argv: ['test_filename']" in handler.error
+
+    doc = Document()
+    handler = CodeHandler(source="""import sys; raise RuntimeError("argv: %r" % sys.argv)""", filename="/test_filename", argv=[10, 20, 30])
+    handler.modify_document(doc)
+
+    assert handler.error is not None
+    assert "argv: ['test_filename', 10, 20, 30]" in handler.error
+
+def test_script_uses_iofuncs():
+    doc = Document()
+    handler = CodeHandler(source=script_uses_iofuncs, filename="/test_filename")
+    handler._logger_text = "%s, %s"
+    handler.modify_document(doc)
+    if handler.failed:
+        raise RuntimeError(handler.error)
+    assert handler.failed is False
+    assert handler.error is None
+    assert len(doc.roots) == 1
+    import sys; sys.stderr.flush()
+
+def test_script_uses_client():
+    doc = Document()
+    handler = CodeHandler(source=script_uses_client, filename="/test_filename")
+    handler._logger_text = "%s, %s"
+    handler.modify_document(doc)
+    if handler.failed:
+        raise RuntimeError(handler.error)
+    assert handler.failed is False
+    assert handler.error is None
+    assert len(doc.roots) == 1
```

Comment:
wow this diff sucks. I just moved all the test methods to be plain funcs and got rid of unneeded `unittest` junk